### PR TITLE
feat(FN-1615): add premium payments table

### PR DESCRIPTION
**Pull request by JoshBinns2000**
## Introduction :pencil2:
We want to be able to view the current fee records associated with a utilisation report in TFM.

## Resolution :heavy_check_mark:
- Updates the return from the `/v1/utilisation-reports/reconciliation-details/:reportId` endpoint
- Adds new types in `libs/common`
- Adds a new method to the fee record entity for converting reported fees to the payment currency
- Adds a new column to the fee record entity

FN-1615-premium-payments-table

### CODE REVIEW

File: trade-finance-manager-ui/component-tests/utilisation-reports/utilisation-report-reconciliation-for-report.component-test.js
Comment (RLavender98) [inline trade-finance-manager-ui/component-tests/utilisation-reports/utilisation-report-reconciliation-for-report.component-test.js:67]: Here can we check for buttons with the correct text rather than just the `data-cy` attribute?

**Review by RLavender98 — CHANGES_REQUESTED** [review]
Looks good, nice job 👍 Mostly fairly minor comments

File: trade-finance-manager-ui/templates/utilisation-reports/_macros/premium-payments-table.njk
Comment (RLavender98) [inline trade-finance-manager-ui/templates/utilisation-reports/_macros/premium-payments-table.njk:28]: I believe we can have the same facility appearing multiple times in a report and therefore multiple times in the table. We might want to surface the fee record id as well but we should leave that to the tickets covering add a payment functionality to decide since it is possibly only of importance to those tickets? And so can perhaps leave this too for now. We probably want to explicitly include duplicates facility ids in some of our test data somewhere, perhaps e2e, but again I think we can safely leave that for now if you agree?

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.ts
Comment (JoshBinns2000) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.ts:7]: I think we should only be removing the type definition from the docstrings and not every `@param` and `@returns` tag (even though in most cases the param names should be self-documenting).

I agree with respect to the rest of the docstring. I will add those details to the type definition instead

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.ts
Comment (JoshBinns2000) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.ts:31]: The difference belongs to a separate ticket I think (I also don't fully know what the difference is, but I think it comes from payments which are made for the same facility in different currencies)

File: libs/common/src/helpers/currency.ts
Comment (JoshBinns2000) [reply libs/common/src/helpers/currency.ts:10]: How about `getFormattedCurrencyAndAmount` to match what we have with `getFormattedReportPeriod`?

File: libs/common/src/helpers/currency.ts
Comment (JoshBinns2000) [reply libs/common/src/helpers/currency.ts:6]: See the comment above about `@param` and `@returns` :)

File: trade-finance-manager-ui/templates/utilisation-reports/_macros/premium-payments-table.njk
Comment (JoshBinns2000) [reply trade-finance-manager-ui/templates/utilisation-reports/_macros/premium-payments-table.njk:28]: I realised this was a problem which is why I added the `loop.index - 1` to `data-cy`, but after reading this I actually think it makes more sense to use the `feeRecordId` as we are likely to want that when we query things in future, so I will update it to use that

File: libs/common/src/utils/monetary-calculation.ts
Comment (JoshBinns2000) [reply libs/common/src/utils/monetary-calculation.ts:11]: Seeing as we are only using the divide one at the minute, I have just replaced this with a single `divideBy` helper. We can add a multiply one in future if we need :)

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/index.test.ts
Comment (JoshBinns2000) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/index.test.ts:94]: The `helpers.test.ts` file provides more explicit checks for how the mapping works with respect to currencies, so I think we can just remove that part of the test name here

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts
Comment (JoshBinns2000) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts:54]: I've updated these tests to try and do this kind of thing - please let me know what you think

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts
Comment (RLavender98) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts:93]: So your test name says you are checking it maps the 'fees paid' to 'reported fees', which is the first assertion, but then you have these two assertions in addition. These should either be removed or if they are important in a different test.

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts
Comment (RLavender98) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts:96]: Can we split this based on the 'and' so one test for how it maps to fees and one test for how it maps to payments? And then we say 'converts' but what is interesting is _how_ it converts so would be good to capture that in the test name.

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts
Comment (RLavender98) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts:66]: I don't think any conversion happens in this case, we map the values like for like no?

File: dtfs-central-api/src/v1/swagger-definitions/curency.js
Comment (RLavender98) [inline dtfs-central-api/src/v1/swagger-definitions/curency.js:1]: typo in file name

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts
Comment (RLavender98) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts:22]: This duplicates some of the assertions in the test below, perhaps we want to just test the base fields that don't have much complexity? I guess here that is the id, facility id, exporter and status

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts
Comment (RLavender98) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts]: Thank you! I'm going to resolve this and leave some more comments as easier than trying to write a reply here

File: libs/common/src/utils/monetary-calculation.ts
Comment (RLavender98) [reply libs/common/src/utils/monetary-calculation.ts]: I still think it is slightly hard for the less mathematically inclined to follow _why_ we divide. I have a maths degree and still it confuses me, I actually liked `applyExchangeRateToAmount` because you didn't need to know what maths it was doing (except it was in the parameters hence the comment). I guess the crucial information here is which way the exchange is going, so it would be good to capture that somehow, perhaps with a parameter names along the lines of `baseCurrencyAmount` and `baseCurrencyToTargetCurrencyExchangeRate`, although perhaps dividing is the other way, 'targetToBase'(?) but I want the function to do that thinking for me 😅 Although I guess `paymentExchangeRate` actually doesn't tell us which way it goes, I'm assuming from what you have written it is  payment currency to fee currency?

**Review by RLavender98 — COMMENTED** [review]
A few comments on the breaking up of tests

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.ts
Comment (RLavender98) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.ts:69]: Should we actually be adding a database column for this, populating on creation and returning the value as part of this PR?

File: dtfs-central-api/src/types/utilisation-reports.ts
Comment (JoshBinns2000) [reply dtfs-central-api/src/types/utilisation-reports.ts:70]: No, but considering that the other fields now have a docstring it feels more consistent to add one for all the other fields too

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts
Comment (RLavender98) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts:23]: I think we can still have one test that tests those four fields that asserts they are mapped over like for like, is there some sort of "object containing" assertion we can do to test just those fields?

The reason for not duplicating is then we have one test in this file fail when the one thing goes wrong, not two, and then we immediately see from the test name of the test that failed exactly what failed, rather than having a rather generic test fail and no understanding from the test name why or what it should actually be doing.

File: dtfs-central-api/src/types/utilisation-reports.ts
Comment (RLavender98) [reply dtfs-central-api/src/types/utilisation-reports.ts:70]: Sorry I meant all the fields, should have dragged down to place the comment on the entire type

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts
Comment (RLavender98) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers.test.ts:23]: I don't really mind too much so will approve even if you leave this as is, but for me a test with a name like this, with this style of assertion for a mapping function should only exist if all the mapping function is doing is mapping each field like for like and doesn't do any other changes no matter the value of the fields.

File: dtfs-central-api/src/types/utilisation-reports.ts
Comment (JoshBinns2000) [reply dtfs-central-api/src/types/utilisation-reports.ts:70]: I think these docstrings make the distinction between `reportedFees` and `reportedPayments` easy to understand. Without it, you have to infer what they mean based on the way the functions are written

File: libs/common/src/utils/monetary-calculation.ts
Comment (JoshBinns2000) [reply libs/common/src/utils/monetary-calculation.ts]: I've added a method to the `FeeRecordEntity` class now rather than having these new helpers, so the way we convert between currencies is now specific to the entity

Comment (ChristianMcc) [reply libs/common/src/utils/monetary-calculation.ts]: Yep I like this, think we should stick with this pattern for doing conversions between currencies where it's not obvious what the conversion is representing, nice

**Review by RLavender98 — APPROVED** [review]
Looks good! Nice improvements on the tests, thank you!

**Pull request merged**
